sklearn/datasets/openml.py

```diff
@@ -263,6 +268,92 @@ def _convert_arff_data(arff_data, col_slice_x, col_slice_y, shape=None):
         raise ValueError('Unexpected Data Type obtained from arff.')
 
 
+def _feature_to_dtype(feature):
+    """Map feature to dtype for pandas DataFrame
+    """
+    if feature['data_type'] == 'string':
+        return object
+    elif feature['data_type'] == 'nominal':
+        return 'category'
+    # only numeric, integer, real are left
+    elif (feature['number_of_missing_values'] != '0' or
+          feature['data_type'] in ['numeric', 'real']):
+        # cast to floats when there are any missing values
+        return np.float64
+    elif feature['data_type'] == 'integer':
+        return np.int64
+    raise ValueError('Unsupported feature: {}'.format(feature))
+
+
+def _chunk_generator(gen, chunksize):
+    """Chunk generator, ``gen`` into lists of length ``chunksize``. The last
+    chunk may have a length less than ``chunksize``."""
+    while True:
+        chunk = list(islice(gen, chunksize))
+        if chunk:
+            yield chunk
+        else:
+            return
+
+
+def _convert_arff_data_dataframe(arrf, columns, features_dict):
+    """Convert the ARFF object into a pandas DataFrame.
+
+    Parameters
+    ----------
+    arrf : dict
+        As obtained from liac-arff object.
+
+    columns : list
+        Columns from dataframe to return.
+
+    features_dict : dict
+        Maps feature name to feature info from openml.
+
+    Returns
+    -------
+    dataframe : pandas DataFrame
+    """
+    pd = check_pandas_support('fetch_openml with as_frame=True')
+
+    attributes = OrderedDict(arrf['attributes'])
+    arrf_columns = list(attributes)
+
+    # calculate chunksize
+    working_memory = get_config()['working_memory']
+    first_row = next(arrf['data'])
+    first_df = pd.DataFrame([first_row], columns=arrf_columns)
+
+    row_bytes = first_df.memory_usage(deep=True).sum()
```

Comment:
This may be more precise than we need, since working_memory only needs to account for *temporary* usage and object-dtype data will still occupy the same heap space when fully constructed. But I suppose if columns_to_keep filters out columns that use a whole lot of heap space then this deep consumption might be relevant

Reply:
No need to fine tune this now in any case

Reply:
But now I'm thinking that actually the working memory we are trying to limit is not the usage of the data frame, but rather the bytes of string space and list storage used by arff.py

Reply:
Fundamentally, using the string space + list storage per line can be used to estimate the `chunksize`. This is using dataframe's `memory_usage` as a shortcut to approximate this.